gax-java/gax/src/main/java/com/google/api/gax/rpc/TransportChannelProvider.java

```diff
@@ -97,6 +97,20 @@ public interface TransportChannelProvider {
    */
   TransportChannelProvider withEndpoint(String endpoint);
 
+  /** True for gRPC transport provider which has no mtlsEndpoint set. */
+  default boolean needsMtlsEndpoint() {
```

Comment:
I know we tried to avoid introducing these new public interfaces by encapsulating the logic to determine if we should use S2A in EndpointContext. Now that we need to have access to mtlsEndpoint in TransportChannelProvider  anyway, maybe we can move [shouldUseS2A](https://github.com/googleapis/sdk-platform-java/blob/f6bdadebb4286c53cbce725dac07bfd2198e4c18/gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java#L318-L344) to InstantiatingGrpcChannelProvider.java? And remove `withUseS2A` from this interface? This could be separate PR.

Reply:
Agreed that by plumbing the MTLS endpoint separately, we no longer need the `shouldUseS2A` logic in EndpointContext. I'll take care of this in a followup PR.

Reply:
@blakeli0 I started working on this and I realized that we should keep [`shouldUseS2A`](https://github.com/googleapis/sdk-platform-java/blob/f6bdadebb4286c53cbce725dac07bfd2198e4c18/gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java#L318-L344) in EndpointContext. The reason being is that we want to skip S2A if using GDCH, or if the client sets a custom endpoint. This info is only available in EndpointContext. 

Given this, we probably want to keep things as is. WDYT?